File: figcli/cmd/app/uninstall/uninstall.go
package uninstall

import (
	"os"
	"os/exec"

	"github.com/spf13/cobra"
)

func NewCommandUninstall() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "uninstall",
		Short: "Uninstall Fig",
		Long:  `Uninstall Fig`,
		Run: func(cmd *cobra.Command, args []string) {
			// Use uninstall script for time being
			installCmd := exec.Command("sh", "-c", "~/.fig/tools/uninstall-script.sh")
			installCmd.Stdout = os.Stdout
			installCmd.Stderr = os.Stderr
			installCmd.Run()

			// // Get user
			// usr, err := user.Current()
			// if err != nil {
			// 	fmt.Println("Error: ", err)
			// 	return
			// }

			// fmt.Println("Deleting .fig folder & completion specs")
			// os.RemoveAll(usr.HomeDir + "/.fig")

			// fmt.Println("Delete backup Fig CLI")
			// os.RemoveAll("/usr/local/bin/fig")

			// fmt.Println("Deleting WKWebViewCache")
			// fig_ipc.RunResetCacheCommand()

			// fmt.Println("Deleting fig defaults & preferences")
			// savedId, err := exec.Command("defaults", "read", "com.mschrage.fig", "uuid").Output()
			// if err != nil {
			// 	fmt.Println("Error: ", err)
			// 	return
			// }

			// exec.Command("defaults", "delete", "com.mschrage.fig").Run()
			// exec.Command("defaults", "delete", "com.mschrage.fig.shared").Run()
			// exec.Command("defaults", "write", "com.mschrage.fig", "uuid", string(savedId)).Run()

			// fmt.Println("Remove iTerm integration (if set up)")
			// os.Remove(usr.HomeDir + "/Library/Application Support/iTerm2/Scripts/AutoLaunch/fig-iterm-integration.py")
			// os.Remove(usr.HomeDir + "/.config/iterm2/AppSupport/Scripts/AutoLaunch/fig-iterm-integration.py")
			// os.Remove(usr.HomeDir + "/Library/Application Support/iTerm2/Scripts/AutoLaunch/fig-iterm-integration.scpt")

			// fmt.Println("Remove VSCode integration (if set up)")
			// files, _ := filepath.Glob(usr.HomeDir + "/.vscode/extensions/withfig.fig-*")
			// for _, file := range files {
			// 	os.RemoveAll(file)
			// }
			// files, _ = filepath.Glob(usr.HomeDir + "/.vscode-insiders/extensions/withfig.fig-*")
			// for _, file := range files {
			// 	os.RemoveAll(file)
			// }

			// fmt.Println("Remove fish integration...")
			// os.Remove(usr.HomeDir + "/.config/fish/conf.d/fig.fish")

			// fmt.Println("Removing fig.sh setup from .profile, .zprofile, .zshrc, .bash_profile, and .bashrc")
			// for _, file := range []string{".profile", ".zprofile", ".zshrc", ".bash_profile", ".bashrc"} {
			// 	if _, err := os.Stat(usr.HomeDir + "/" + file); err == nil {
			// 		// File exists
			// 		lines, err := os.ReadFile(usr.HomeDir + "/" + file)
			// 		if err != nil {
			// 			fmt.Println("Error: ", err)
			// 			return
			// 		}

			// 		lines = bytes.Replace(lines, []byte("#### FIG ENV VARIABLES ####\n"), []byte(""), -1)
			// 		lines = bytes.Replace(lines, []byte("# Please make sure this block is at the start of this file.\n"), []byte(""), -1)
			// 		lines = bytes.Replace(lines, []byte("[ -s ~/.fig/shell/pre.sh ] && source ~/.fig/shell/pre.sh\n"), []byte(""), -1)
			// 		lines = bytes.Replace(lines, []byte("[ -s ~/.fig/fig.sh ] && source ~/.fig/fig.sh\n"), []byte(""), -1)
			// 		lines = bytes.Replace(lines, []byte("# Please make sure this block is at the end of this file.\n"), []byte(""), -1)
			// 		lines = bytes.Replace(lines, []byte("#### END FIG ENV VARIABLES ####\n"), []byte(""), -1)

			// 		os.WriteFile(usr.HomeDir+"/"+file, lines, 0755)
			// 	}
			// }

			// fmt.Println("Removing fish integration")
			// if _, err := os.Stat(usr.HomeDir + "/.config/fish/config.fish"); err == nil {
			// 	// File exists
			// 	lines, err := os.ReadFile(usr.HomeDir + "/.config/fish/config.fish")
			// 	if err != nil {
			// 		fmt.Println("Error: ", err)
			// 		return
			// 	}

			// 	fishInstall := fmt.Sprintf("contains %s/.fig/bin %s; or set -Ua fish_user_paths %s/.fig/bin\n",
			// 		usr.HomeDir, os.Getenv("fish_user_paths"), usr.HomeDir)

			// 	lines = bytes.Replace(lines, []byte(fishInstall), []byte(""), -1)

			// 	os.WriteFile(usr.HomeDir+"/.config/fish/config.fish", lines, 0755)
			// }
			// os.Remove(usr.HomeDir + "/.config/fish/conf.d/fig.fish")

			// fmt.Println("Remove Hyper plugin, if it exists")
			// if _, err := os.Stat(usr.HomeDir + "/.hyper.js"); err == nil {
			// 	lines, err := os.ReadFile(usr.HomeDir + "/.hyper.js")
			// 	if err != nil {
			// 		fmt.Println("Error: ", err)
			// 		return
			// 	}

			// 	lines = bytes.Replace(lines, []byte("\"fig-hyper-integration\","), []byte(""), -1)
			// 	lines = bytes.Replace(lines, []byte("\"fig-hyper-integration\""), []byte(""), -1)

			// 	os.WriteFile(usr.HomeDir+"/.hyper.js", lines, 0755)
			// }

			// fmt.Println("Finished removing fig resources. You may now delete the Fig app by moving it to the Trash.")

			// os.RemoveAll(usr.HomeDir + "/Library/Input Methods/FigInputMethod.app")
			// os.RemoveAll(usr.HomeDir + "/Applications/Fig.app")

			// fig_ipc.QuitCommand()
		},
	}

	return cmd
}

File: fig_cli/src/cli/app/uninstall.rs
use std::path::{
    Path,
    PathBuf,
};

use clap::Args;
use fig_directories::home_dir;
use fig_telemetry::{
    TrackEvent,
    TrackSource,
};
use tokio::io::{
    AsyncReadExt,
    AsyncWriteExt,
};
use tracing::warn;

use crate::cli::installation::{
    uninstall_cli,
    InstallComponents,
};
use crate::daemon::IS_RUNNING_DAEMON;

async fn remove_in_dir_with_prefix_unless(dir: &Path, prefix: &str, unless: impl Fn(&str) -> bool) {
    if let Ok(mut entries) = tokio::fs::read_dir(dir).await {
        while let Ok(Some(entry)) = entries.next_entry().await {
            if let Some(name) = entry.file_name().to_str() {
                if name.starts_with(prefix) && !unless(name) {
                    tokio::fs::remove_file(entry.path()).await.ok();
                    tokio::fs::remove_dir_all(entry.path()).await.ok();
                }
            }
        }
    }
}

#[derive(Debug, Args)]
pub struct UninstallArgs {
    /// Remove configuration and data files
    #[clap(long, action)]
    pub user_data: bool,
    /// Remove executable and
    #[clap(long, action)]
    pub app_bundle: bool,
    /// Remove input method
    #[clap(long, action)]
    pub input_method: bool,
    /// Remove terminal integrations (i.e. VSCode, iTerm2, etc.)
    #[clap(long, action)]
    pub terminal_integrations: bool,
    /// Remove Fig daemon
    #[clap(long, action)]
    pub daemon: bool,
    /// Remove dotfile shell integration
    #[clap(long, action)]
    pub dotfiles: bool,
    /// Remove SSH integration
    #[clap(long, action)]
    pub ssh: bool,
    /// Do not open the uninstallation page
    #[clap(long, action)]
    pub no_open: bool,
}

pub async fn uninstall_mac_app(uninstall_args: &UninstallArgs) {
    // Send uninstall telemetry event
    let tel_join = tokio::task::spawn(async move {
        fig_telemetry::emit_track(
            TrackEvent::UninstallApp,
            if *IS_RUNNING_DAEMON.lock() {
                TrackSource::Daemon
            } else {
                TrackSource::Cli
            },
            [("source", "fig app uninstall")],
        )
        .await
        .ok();
    });

    if !uninstall_args.no_open {
        // Open the uninstallation page
        let email = fig_auth::get_email().unwrap_or_default();
        let version = fig_auth::get_default("versionAtPreviousLaunch").unwrap_or_default();
        fig_util::open_url(format!("https://fig.io/uninstall?email={email}&version={version}",)).ok();
    }

    if uninstall_args.app_bundle {
        uninstall_app_bundle().await;
    }

    if uninstall_args.user_data {
        uninstall_user_data().await;
    }

    if uninstall_args.input_method {
        uninstall_input_method().await;
    }

    if uninstall_args.terminal_integrations {
        uninstall_terminal_integrations().await;
    }

    if uninstall_args.dotfiles {
        uninstall_dotfiles();
    }

    if uninstall_args.ssh {
        uninstall_ssh();
    }

    // Daemon must come last
    if uninstall_args.daemon {
        uninstall_daemon();
    }

    tel_join.await.ok();
}

async fn uninstall_app_bundle() {
    let app_path = PathBuf::from("Applications").join("Fig.app");
    if app_path.exists() {
        tokio::fs::remove_dir_all(&app_path)
            .await
            .map_err(|err| warn!("Failed to remove Fig.app: {err}"))
            .ok();
    }

    // Remove launch agents
    if let Some(home) = home_dir() {
        let launch_agents = home.join("Library").join("LaunchAgents");
        remove_in_dir_with_prefix_unless(&launch_agents, "io.fig.", |p| p.contains("daemon")).await;
    } else {
        warn!("Could not find home directory");
    }

    if let Err(err) = uninstall_cli(InstallComponents::BINARY) {
        warn!("Could not uninstall CLI: {err}");
    }
}

async fn uninstall_user_data() {
    // Delete Fig defaults on macOS
    #[cfg(target_os = "macos")]
    {
        let uuid = fig_auth::get_default("uuid").unwrap_or_default();
        tokio::process::Command::new("defaults")
            .args(["delete", "com.mschrage.fig"])
            .output()
            .await
            .map_err(|err| warn!("Failed to delete defaults: {err}"))
            .ok();
        tokio::process::Command::new("defaults")
            .args(["delete", "com.mschrage.fig.shared"])
            .output()
            .await
            .map_err(|err| warn!("Failed to delete defaults: {err}"))
            .ok();
        tokio::process::Command::new("defaults")
            .args(["write", "com.mschrage.fig", "uuid", &uuid])
            .output()
            .await
            .map_err(|err| warn!("Failed to write defaults: {err}"))
            .ok();
    }

    // Delete data dir
    if let Some(fig_data_dir) = fig_directories::fig_data_dir() {
        tokio::fs::remove_dir_all(&fig_data_dir)
            .await
            .map_err(|err| warn!("Could not remove {}: {err}", fig_data_dir.display()))
            .ok();
    } else {
        warn!("Could not find data directory");
    }

    // Delete the ~/.fig folder
    if let Some(fig_dir) = fig_directories::fig_dir() {
        tokio::fs::remove_dir_all(fig_dir)
            .await
            .map_err(|err| warn!("Could not remove ~/.fig folder: {err}"))
            .ok();
    } else {
        warn!("Could not find .fig folder");
    }
}

async fn uninstall_input_method() {
    if let Some(home) = fig_directories::home_dir() {
        // Remove the app
        let fig_input_method_app = home.join("Library").join("Input Methods").join("FigInputMethod.app");

        if fig_input_method_app.exists() {
            tokio::fs::remove_dir_all(fig_input_method_app)
                .await
                .map_err(|err| warn!("Could not remove FigInputMethod.app: {err}"))
                .ok();
        }
    } else {
        warn!("Could not find home directory");
    }
}

async fn uninstall_terminal_integrations() {
    // Delete integrations
    if let Some(home) = home_dir() {
        // Delete iTerm integratione
        for path in &[
            "Library/Application Support/iTerm2/Scripts/AutoLaunch/fig-iterm-integration.py",
            ".config/iterm2/AppSupport/Scripts/AutoLaunch/fig-iterm-integration.py",
            "Library/Application Support/iTerm2/Scripts/AutoLaunch/fig-iterm-integration.scpt",
        ] {
            tokio::fs::remove_file(home.join(path))
                .await
                .map_err(|err| warn!("Could not remove iTerm integration {path}: {err}"))
                .ok();
        }

        // Delete VSCode integration
        for (folder, prefix) in &[
            (".vscode/extensions", "withfig.fig-"),
            (".vscode-insiders/extensions", "withfig.fig-"),
            (".vscode-oss/extensions", "withfig.fig-"),
        ] {
            let folder = home.join(folder);
            remove_in_dir_with_prefix_unless(&folder, prefix, |_| false).await;
        }

        // Remove Hyper integration
        let hyper_path = home.join(".hyper.js");
        if hyper_path.exists() {
            // Read the config file
            match tokio::fs::File::open(&hyper_path).await {
                Ok(mut file) => {
                    let mut contents = String::new();
                    match file.read_to_string(&mut contents).await {
                        Ok(_) => {
                            contents = contents.replace("\"fig-hyper-integration\",", "");
                            contents = contents.replace("\"fig-hyper-integration\"", "");

                            // Write the config file
                            match tokio::fs::File::create(&hyper_path).await {
                                Ok(mut file) => {
                                    file.write_all(contents.as_bytes())
                                        .await
                                        .map_err(|err| warn!("Could not write to Hyper config: {err}"))
                                        .ok();
                                },
                                Err(err) => {
                                    warn!("Could not create Hyper config: {err}")
                                },
                            }
                        },
                        Err(err) => {
                            warn!("Could not read Hyper config: {err}");
                        },
                    }
                },
                Err(err) => {
                    warn!("Could not open Hyper config: {err}");
                },
            }
        }

        // Remove Kitty integration
        let kitty_path = home.join(".config").join("kitty").join("kitty.conf");
        if kitty_path.exists() {
            // Read the config file
            match tokio::fs::File::open(&kitty_path).await {
                Ok(mut file) => {
                    let mut contents = String::new();
                    match file.read_to_string(&mut contents).await {
                        Ok(_) => {
                            contents = contents.replace("watcher ${HOME}/.fig/tools/kitty-integration.py", "");
                            // Write the config file
                            match tokio::fs::File::create(&kitty_path).await {
                                Ok(mut file) => {
                                    file.write_all(contents.as_bytes())
                                        .await
                                        .map_err(|err| warn!("Could not write to Kitty config: {err}"))
                                        .ok();
                                },
                                Err(err) => {
                                    warn!("Could not create Kitty config: {err}")
                                },
                            }
                        },
                        Err(err) => {
                            warn!("Could not read Kitty config: {err}");
                        },
                    }
                },
                Err(err) => {
                    warn!("Could not open Kitty config: {err}");
                },
            }
        }
        // TODO: Add Jetbrains integration
    }
}

fn uninstall_daemon() {
    uninstall_cli(InstallComponents::DAEMON)
        .map_err(|err| warn!("Could not uninstall daemon: {err}"))
        .ok();
}

fn uninstall_dotfiles() {
    uninstall_cli(InstallComponents::DOTFILES)
        .map_err(|err| warn!("Could not uninstall dotfiles: {err}"))
        .ok();
}

fn uninstall_ssh() {
    uninstall_cli(InstallComponents::SSH)
        .map_err(|err| warn!("Could not uninstall SSH: {err}"))
        .ok();
}
